src/Eryph.GenePool.Packing/GenePackerStream.cs

```diff
@@ -0,0 +1,233 @@
+    }
+
+    public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+    {
+        return WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+    }
+
+    public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+    {
+        if (_isDisposed)
+            throw new ObjectDisposedException(GetType().Name);
+
+        if (buffer.Length == 0)
+            return;
+
+        var currentBuffer = buffer;
+
+        while ((_currentChunk?.Length ?? 0) + currentBuffer.Length > _chunkSize)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            var bytesToWrite = (int)(_chunkSize - (_currentChunk?.Length ?? 0));
+            await WriteChunkAsync(currentBuffer[..bytesToWrite], cancellationToken).ConfigureAwait(false);
+            await EndChunkAsync().ConfigureAwait(false);
+            StartChunk();
+            currentBuffer = currentBuffer[bytesToWrite..];
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+        await WriteChunkAsync(currentBuffer, cancellationToken).ConfigureAwait(false);
+    }
+
+    private async ValueTask WriteChunkAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+    {
+        if (_currentChunk is null)
+            StartChunk();
+
+        await _currentChunk.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);
+        _incrementalHash.AppendData(buffer.Span);
+        _totalBytes += buffer.Length;
+    }
+
+    public override bool CanRead => false;
+
+    public override bool CanSeek => false;
+
+    public override bool CanWrite => true;
+
+    public override long Length => _totalBytes;
+
+    public override long Position
+    {
+        get => throw new NotSupportedException();
+        set => throw new NotSupportedException();
+    }
+
+    [MemberNotNull(nameof(_currentChunk))]
+    private void StartChunk()
+    {
+        var chunkPath = Path.Combine(_chunksDirectory.FullName, $"{_chunks.Count}.part");
+        _currentChunk = new FileStream(chunkPath, FileMode.Create, FileAccess.Write, FileShare.None);
+    }
+
+    private async ValueTask EndChunkAsync()
+    {
+        if (_currentChunk is null)
+            throw new InvalidOperationException("Chunk is not started");
+
+        var chunkPath = _currentChunk.Name;
+        await _currentChunk.DisposeAsync().ConfigureAwait(false);
+        _currentChunk = null;
+
+        EndChunk(chunkPath);
+    }
+
+    private void EndChunk()
+    {
+        if (_currentChunk is null)
+            throw new InvalidOperationException("Chunk is not started");
+
+        var chunkPath = _currentChunk.Name;
+        _currentChunk.Dispose();
+        _currentChunk = null!;
+
+        EndChunk(chunkPath);
+    }
+
+    private void EndChunk(string chunkPath)
+    {
+        Span<byte> span = stackalloc byte[_incrementalHash.HashLengthInBytes];
```

Comment:
Most MS code I have seen last years is using ArrayPool<T>.Shared.Rent, looks to be a bit faster:
https://stackoverflow.com/questions/55229518/why-allocation-on-arraypool-is-faster-then-allocation-on-stack

Reply:
Something additional just seen:

``` shell
packing ubuntu-23.04 to dbosoft/ubuntu-23.04/20240204
{
   "geneset": "dbosoft/ubuntu-23.04/20240204"
}Preparing catlet...
Catlet prepared
```

There is no line wrap behind the json output.